### PR TITLE
Remove global version of appEventBus

### LINKED ISSUE #4888: appEventBus

### PR DESCRIPTION
**Pull request by Nokel81**
Signed-off-by: Sebastian Malton <sebastian@malton.name>

### CODE REVIEW

File: src/common/utils/async-result.ts [hunk 1]
```diff
@@ -3,5 +3,9 @@
  * Licensed under MIT License. See LICENSE in root directory for more information.
  */
 export type AsyncResult<Response, Error = string> =
-  | { callWasSuccessful: true; response: Response }
+  | (
+    Response extends void
+    ? { callWasSuccessful: true; response?: undefined }
+    : { callWasSuccessful: true; response: Response }
+  )
```
Comment (jansav) [inline src/common/utils/async-result.ts:10]: This doesn't seem right. If the expected success response is undefined, then it should be used as a `Response` parameter. Now this means that we have to always make sure that `response` exists, which isn't fine.

File: src/main/helm/exec-helm/exec-helm.injectable.ts [hunk 2]
```diff
@@ -16,20 +16,9 @@ const execHelmInjectable = getInjectable({
...
-      }
-
-      return {
-        callWasSuccessful: false,
-        error: response.error.stderr || response.error.error.message,
-      };
-    };
+    return async (args) => execFile(helmBinaryPath, args, {
+      maxBuffer: 32 * 1024 * 1024 * 1024, // 32 MiB
+    });
```
Comment (jansav) [inline src/main/helm/exec-helm/exec-helm.injectable.ts:22]: Why do we want to expose complexity of `stderr` to usages?

File: src/main/helm/repositories/add-helm-repository/add-helm-repository.injectable.ts [hunk 3]
```diff
@@ -54,7 +55,18 @@ const addHelmRepositoryInjectable = getInjectable({
...
+
+      if (result.callWasSuccessful) {
+        return {
+          callWasSuccessful: true,
+        };
+      }
+
+      return {
+        callWasSuccessful: false,
+        error: result.error.stderr || result.error.error.message,
```
Comment (jansav) [inline src/main/helm/repositories/add-helm-repository/add-helm-repository.injectable.ts]: This is the type of complexity that I mean with earlier comment about `stderr`.

File: src/common/cluster-store/cluster-store.ts [hunk 4]
```diff
@@ -115,7 +116,7 @@ export class ClusterStore extends BaseStore<ClusterStoreModel> {
   }
 
   addCluster(clusterOrModel: ClusterModel | Cluster): Cluster {
-    appEventBus.emit({ name: "cluster", action: "add" });
+    this.dependencies.emitAppEvent({ name: "cluster", action: "add" });
```
Comment (Iku-turso) [inline src/common/cluster-store/cluster-store.ts:119]: I'm thinking `emitAppEvent` is too low an abstraction for this context. It could instead be `emitClusterAdd()` or maybe even `emit<ClusterEvent>("add")`. The former would replace the logical dependency with a more trackable one. Eg. we could discover listeners of `ClusterEvent` more easily (if even present in this codebase). Also it could add type safety to the payload of the event, ie. `action: "add"`.

File: src/main/resource-applier/resource-applier.ts [hunk 5]
```diff
@@ -49,23 +57,17 @@ export class ResourceApplier {
...
-        throw error.stderr ?? error;
-      }
+    const result = await this.dependencies.execFile(kubectlPath, args);
 
-      throw error;
+    if (result.callWasSuccessful) {
+      return result.response;
     }
+
+    throw result.error.stderr || result.error.error;
```
Comment (jansav) [inline src/main/resource-applier/resource-applier.ts]: Again, this doesn't make sense. `execFile` should abstract this type of complexity away from usages. Best would be that we define error as an interface, e.g.

```typescript 
interface Error { 
  code: string;
  message: string;
  
  original: unknown // Error, string, whatever
} 
```

In this way, we could make `AsyncResult`

```typescript
type AsyncResult<Response> =  { callWasSuccessful: true, response: Response } | { callWasSuccessful: false, error: Error }
```

File: src/common/utils/async-result.ts (hunk 1 above)
Comment (jansav) [reply src/common/utils/async-result.ts:10]: Read: https://github.com/lensapp/lens/pull/6096/files#r1000512666

File: src/main/router/router.test.ts [hunk 6]
```diff
@@ -32,6 +33,7 @@ describe("router", () => {
     const di = getDiForUnitTesting({ doGeneralOverrides: true });
 
     mockFs();
+    di.permitSideEffects(fsInjectable);
```
Comment (jansav) [inline src/main/router/router.test.ts:36]: Why is this needed?

File: src/main/electron-app/runnables/setup-ipc-main-handlers/setup-ipc-main-handlers.ts [hunk 7]
```diff
@@ -78,11 +82,11 @@ export const setupIpcMainHandlers = ({
   });
 
   ipcMainHandle(clusterKubectlApplyAllHandler, async (event, clusterId: ClusterId, resources: string[], extraArgs: string[]) => {
-    appEventBus.emit({ name: "cluster", action: "kubectl-apply-all" });
+    emitAppEvent({ name: "cluster", action: "kubectl-apply-all" });
     const cluster = ClusterStore.getInstance().getById(clusterId);
 
     if (cluster) {
-      const applier = new ResourceApplier(cluster);
+      const applier = createResourceApplier(cluster);
```
Comment (Iku-turso) [inline src/main/electron-app/runnables/setup-ipc-main-handlers/setup-ipc-main-handlers.ts:92]: As common practice, it could be better to not inject factory functions, if an already "factored" instance can be injected instead. Here: `applier` could be the dependency instead of `createResourceApplier`.

File: src/main/helm/repositories/add-helm-repository/add-helm-repository.injectable.ts (hunk 3 above)
Comment (Iku-turso) [inline src/main/helm/repositories/add-helm-repository/add-helm-repository.injectable.ts]: I'm seeing this `error.error` duplicated in many places. The funk is a bit too much. Improve naming? One could be "errorObject"? Or the root error could be "rejection"?

File: src/common/utils/async-result.ts (hunk 1 above)
Comment (jansav) [reply src/common/utils/async-result.ts:10]: > This doesn't seem right. If the expected success response is undefined, then it should be used as a Response parameter. Now this means that we have to always make sure that response exists, which isn't fine.


Misread the code, it seems fine.

**Review by Iku-turso — CHANGES_REQUESTED** [review]
Check out comments from me and @jansav.

File: src/main/router/router.test.ts (hunk 6 above)
Comment (Nokel81) [reply src/main/router/router.test.ts:36]: Because the router tests still use `mockFs`

File: src/main/electron-app/runnables/setup-ipc-main-handlers/setup-ipc-main-handlers.ts (hunk 7 above)
Comment (Nokel81) [reply src/main/electron-app/runnables/setup-ipc-main-handlers/setup-ipc-main-handlers.ts:92]: small difference

File: src/main/helm/repositories/add-helm-repository/add-helm-repository.injectable.ts (hunk 3 above)
Comment (Nokel81) [reply src/main/helm/repositories/add-helm-repository/add-helm-repository.injectable.ts]: Or maybe in this case we could just wrap it better.

File: src/common/cluster-store/cluster-store.ts (hunk 4 above)
Comment (Iku-turso) [reply src/common/cluster-store/cluster-store.ts:119]: Resolved for being just food for thought.

File: src/main/electron-app/runnables/setup-ipc-main-handlers/setup-ipc-main-handlers.ts (hunk 7 above)
Comment (Iku-turso) [reply src/main/electron-app/runnables/setup-ipc-main-handlers/setup-ipc-main-handlers.ts:92]: Sometimes, sometimes not :) Resolved for already communicating the general point somewhere else.

File: src/main/router/router.test.ts (hunk 6 above)
Comment (Iku-turso) [reply src/main/router/router.test.ts:36]: Answer's good, resolved.

**Review by Iku-turso — APPROVED** [review]
